docs/do-notation.md

```diff
@@ -0,0 +1,79 @@
+# Do Notation
+
+Both [Haskell](https://wiki.haskell.org/Monad#do-notation)
+and [PureScript](https://github.com/purescript/documentation/blob/master/language/Syntax.md#do-notation)
+languages provide syntactic sugar for working with monads in the form of do notation.
+
+`fp-ts` provides it's own implementation of do notation which can help to simplify effectful
+code.
+
+Let's take a look at an example of how do notation can help to simplify our code. Here we have
+a bit of code which reads two values from the command line, prints them and stores them
+in an object with `x` and `y` properties.
+
+```ts
+import { pipe } from 'fp-ts/function'
+import * as T from 'fp-ts/Task'
+
+declare const print: (s: string) => T.Task<void>
+declare const readLine: T.Task<string>
```

Comment:
Since `readLine` is already a Task and not a lazy function returning a Task, I think the `apS` method is more appropriate here:
```
const mainDo: T.Task<{ x: string, y: string }> = pipe(
  T.Do,
  T.apS('x', readLine),
  T.apS('y', readLine),
  T.chainFirst(({ x }) => print(x)),
  T.chainFirst(({ y }) => print(y)),
)
```
I'm not sure how to introduce both concepts but maybe we can update your example with `readline` being:
```
declare const readLine: () => T.Task<string>
```
and the bind being used like: 
```
T.bind('x', () => readLine()),
// or even:
T.bind('x', readLine),
```
and below, we could add that if you have `Tasks` not wrapped in a lazy function, you can use `apS`.
What do you think?

Reply:
In the case of `Task`, you can use the default exported `apS` when sequentiality doesn't matter. 

For example while running the following program:

```ts
const parallel: T.Task<{ x: string, y: string }> = pipe(
  T.Do,
  T.apS('x', readLine),
  T.apS('y', readLine)
)
```

`y` **may** be requested before `x`. However in the following program:

```ts
const sequential: T.Task<{ x: string, y: string }> = pipe(
  T.Do,
  T.apS('x', readLine),
  T.bind('y', () => readLine)
)
```

`y` is guaranteed being requested after `x`.

Reply:
Interesting, thanks for the explanation.

I think that sticking with `bind` for the initial example is the best as it's the closest to the do notation from Haskell/Purescript and so serves as a direct translation.

The idea is to tack on a bunch of other examples at the end of this page so I think that would be a good place to add examples with `apS`, `bindTo` etc.

@pfgray tagging you as well for visibility

Reply:
I don't know if it's worth mentioning on the initial example, but another big difference between `bind` and `apS` is that you can use what's currently in the Monad to compute the value with `bind` while you can't with `apS`.
Here's an example:
```ts
import { pipe } from 'fp-ts/function'
import * as T from 'fp-ts/Task'

declare const print: (s: string) => T.Task<void>
declare const readLine: T.Task<string>
declare const formatString: (s: string) => T.Task<string>

const mainDo: T.Task<{ x: string, y: string }> = pipe(
  T.Do,
  T.bind('x', () => readLine),
  T.bind('y', ({ x }) => formatString(x)),
  T.chainFirst(({ x }) => print(x)),
  T.chainFirst(({ y }) => print(y)),
)
```
We can put in with the other examples at the end the page anyway.